keps/sig-architecture/4402-go-workspaces/README.md

```diff
@@ -0,0 +1,437 @@
+implementation.
+
+However, k8s.io/code-generator is more challenging.  It holds all of our tools
+and some scripts for invoking them, and is used by a non-trivial number of
+downstream projects (operators, custom API servers, etc).  Making a v2 of this
+(which itself lives in staging, and is subject to this KEP!!) is something we
+have not done before.  The simple answer is to rely on the fact that it was
+never really versioned before, and just make the changes.  We could move it all
+to a new "legacy-code-generator" staging repo and explain that it gets no
+further support, or we could just tag the last legacy release of
+k8s.io/code-generator and tell people to sync to that tag.  Same effect.
+
+Another risk is that this depends on
+[Go 1.22](https://tip.golang.org/doc/go1.22) and
+[vendoring in workspace mode](https://github.com/golang/go/issues/60056), which
+are not yet released. Go 1.22 is scheduled to release in February 2024, which
+seems like good timing for k8s 1.30, but has not yet been confirmed as
+plan-of-record.  As of Jan 13, 2024, people on the release team think it's
+likely that we will release k8s 1.30 on Go 1.22, but
+[that work](https://github.com/kubernetes/release/issues/3280) has not yet
+begun.  The closer we get to the Kubernetes release, the more risk this
+introduces, so if we can't get it in early, we should probably wait until 1.31.
+
+Another risk is that there are many tools in use which DO NOT HAVE TESTS.  They
+seem to work today, but upon closer inspection it's not always clear that they
+work correctly or completely, or what the exact semantics are intended to be.
+Changing these is dangerous.  As we proceed with this work, we will have to add
+tests to tools and verifiers, and look closely at the results of the various
+code-generators.  Ideally there will be NO CHANGE in generated code.
+
+### Test Plan
+
+We will add some tests to untested tools, though complete functional coverage
+is difficult, and would be a prohibitively high bar (e.g. it's unreasonable to
+demand net-new tests for every codegen tool which exercises every path). That
+said, we will look carefully at how each tool works, to mitigate risks.  At the
+time of this writing, the author has spent upwards of 100 hours in the
+debugger, single-stepping through each tool.
+
+[X] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+We will add tests for critical subsystems and tools being changed (e.g. gengo's
+package loading).
+
+##### Integration tests
+
+N/A - see e2e
+
+##### e2e tests
+
+The best test for work like this is, more or less, "did it work?".
+
+If the codegen tools emit the exact same code (or changes are manually
+verifiable as correct) then "it works".  Part of this process must be to
```

Comment:
Is "it works" defined as just building and passing the k8s test suite? Or would the output need to also be identical? Is that even possible?